### ISSUE TITLE
Proposal: Method to provide different elements to initiate Mouse vs Touch dragging

### ISSUE THREAD

**Issue by thesunny** [LouisBrunner/dnd-multi-backend#41]
When we drag using the mouse, the expectation is that we can click and drag anywhere on the element to initiate the drag. Consider a file explorer in a code editor for example. We can usually click and drag the filename to move it.

When we drag using touch, the expectation is that we click on a drag handle. The reason why is because we need a way to scroll and usually that means dragging anywhere except on the drag handle. Consider how we reorganize bookmarks on an iPhone. We initiate the move by dragging on the "three bars" icon on the right.

From the current API, it does not appear this is possible (or is not documented).

My current scenario is that I have a tree structure with drag bars on the right. I want to allow mouse drags to start by dragging the tree items (they can also drag the drag bars but it is not necessary). For touch, I only want to initiate dragging by dragging the bars.

**Comment by LouisBrunner** [comment]
Hi @thesunny,

This isn't possible at the moment, but that is a really interesting proposal! Do you have any ideas about how the API/usage would look like? I am especially interested about how to assign a particular backend to a particular source/target.

Here is my first draft (feel free to add your own!), would it fit your purpose?

```
import { useMultiDrag } from 'react-dnd-multi-backend'

function ItemList(props) {
  const [collectedProps, drag] = useMultiDrag({
    item: { id, type },
  }, (backend) => (backend.id === 'mouse'))
  return <div ref={drag}>...</div>
}

...

const HTML5toTouch = {
  backends: [
    {
      id: 'html5',
      backend: HTML5Backend,
      transition: MouseTransition
      // by default, will dispatch a duplicate `mousedown` event when this backend is activated
    },
    {
      id: 'touch',
      backend: TouchBackend,
      // Note that you can call your backends with options
      options: {enableMouseEvents: true},
      preview: true,
      transition: TouchTransition,
      // will not dispatch a duplicate `touchstart` event when this backend is activated
      skipDispatchOnTransition: true
    }
  ]
};
```

**Comment by thesunny** [comment]
Hi @LouisBrunner,

Thanks for responding and your sample code.

Here's one approach that would work for me and would be compatible with react-dnd API (with one extra options Object):

```js
import { useMultiDrag } from 'react-dnd-multi-backend'

function ItemList(props) {
  const [collectedProps, drag, preview, {html5Ref, touchRef}] = useMultiDrag({
    item: { id, type },
  })
  return <div ref={html5Ref}>I'm the item <span ref={touchRef}><HandleIcon /></span></div>
}

const HTML5toTouch = {
  backends: [
    {
      handleKey: 'html5Ref',
      backend: HTML5Backend,
      transition: MouseTransition
      // by default, will dispatch a duplicate `mousedown` event when this backend is activated
    },
    {
      handleKey: 'touchRef',
      backend: TouchBackend,
      // Note that you can call your backends with options
      options: {enableMouseEvents: true},
      preview: true,
      transition: TouchTransition,
      // will not dispatch a duplicate `touchstart` event when this backend is activated
      skipDispatchOnTransition: true
    }
  ]
};
```

This would handle the use case of being able to drag anywhere on the item with the mouse and being able to drag the handle for touch.

Long-term, I wonder if "Dnd Multi backend" would better be served as a library that is specifically focused on merging HTML5 drag and touch drag together because I can't actually think of any other uses case than HTML5 and touch.

**Comment by LouisBrunner** [comment]
Thanks for your sample @thesunny, I really like the idea of returning the refs in an object and the preview as well!

I will start working on a basic version in the following days.

**Comment by thesunny** [comment]
That sounds great.

I wasn't expecting it to be considered so quickly. You've made me glad to have taken the time to write the proposal.

Thank you!

**Comment by LouisBrunner** [comment]
@thesunny,

I am very sorry about the delay but I finally got to finish a first draft for that proposal.

I published it under the `next` tag of `npm` (version `5.1.0-rc01`) and you can see the documentation on the [`next` branch](https://github.com/LouisBrunner/dnd-multi-backend/tree/next) on GitHub.

Any feedback would be greatly appreciated!

**Comment by thesunny** [comment]
Hey thanks @LouisBrunner, I will try to take a look on Wednesday.

Putting it in my calendar. :)

**Comment by LouisBrunner** [comment]
@thesunny,

Unfortunately, due to messing up my npm versions, I upgraded the package to `6.0.0` and the draft with your proposal is now under `6.1.0-rc01`.

Sorry about the confusion!

**Comment by thesunny** [comment]
Thanks for the update. Looking forward to checking it out tomorrow.

**Comment by thesunny** [comment]
Hi @LouisBrunner ,

Thanks for working on this new feature! It will be a great addition to your library.

During the process of evaluating `6.1.0-rc01`, this issue came up.

Importing `react-dnd-multi-backend` works fine. When importing at the path for `HTML5toTouch` using the instructions in the documentation https://github.com/LouisBrunner/dnd-multi-backend/tree/next/packages/react-dnd-multi-backend#usage like this:

```js
import HTML5toTouch from "react-dnd-multi-backend/dist/esm/HTML5toTouch" // or any other pipeline
```

This error appears:

![image](https://user-images.githubusercontent.com/469587/81219099-1ba5bd80-8f94-11ea-9acc-06373f3669e0.png)

These are my package versions

```
    ...
    "react-dnd": "^10.0.2",
    "react-dnd-html5-backend": "^10.0.2",
    "react-dnd-multi-backend": "6.1.0-rc01",
    "react-dnd-touch-backend": "^10.0.2",
    ...
```

The error doesn't occur when importing `react-dnd-multi-backend`. It's only when importing at the `HTML5toTouch` path.

FYI, I'm using it with `nextjs`

**Comment by LouisBrunner** [comment]
It depends on your build system, you might need to do `import HTML5toTouch from "react-dnd-multi-backend/dist/cjs/HTML5toTouch"` (see here https://github.com/LouisBrunner/dnd-multi-backend/tree/master/packages/react-dnd-multi-backend#npm-installation).

**Comment by thesunny** [comment]
Thanks for the clarification. It felt like a WebPack/Babel/Compilation issue.

May I recommend that the example have a comment right in it that it could be `/cjs/` depending on the build system. If I wasn't testing, I may have moved on to another library.

**Comment by thesunny** [comment]
Found a documentation issue.

```js
  const [[dragProps], {html5: [html5Props, html5Drag], touch: [touchProps, touchDrag]}] = useMultiDrag({
    item: {type: 'card', color: props.color},
    collect: (monitor) => {
      return {
        isDragging: monitor.isDragging(),
      };
    },
  });
```

Under the `html5` and `touch` property values, there is an array. The first argument are `props` and not the `ref`.

**Comment by thesunny** [comment]
Drag and Drop works with a mouse; however, it only drags the handle and not the container. My instinct was that we define the container by using the `ref` returns as part of the first argument; however, when I drag, only the handle moves.

**Comment by thesunny** [comment]
Drag and Drop with touch doesn't seem to work for me, either on my device or on the Chrome simulator; however, I may have made an error in the code so I post it here in case you have any ideas on what's wrong.

```js
import Head from "next/head"
import { useState } from "react"
import { useMultiDrag } from "react-dnd-multi-backend"
import { DndProvider } from "react-dnd-multi-backend"
import HTML5toTouch from "react-dnd-multi-backend/dist/cjs/HTML5toTouch" // or any other pipeline

const initialItems = [
  { id: 1, title: "Alpha" },
  { id: 2, title: "Bravo" },
  { id: 3, title: "Charlie" },
  { id: 4, title: "Delta" },
  { id: 5, title: "Echo" },
]

const Page = () => {
  return (
    <div className="container">
      <Head>
        <link
          rel="stylesheet"
          href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css"
          integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh"
          crossOrigin="anonymous"
        />
      </Head>
      <h1>React DND Demo</h1>
      <DndProvider options={HTML5toTouch}>
        <DragAndDrop />
      </DndProvider>
    </div>
  )
}

const DragAndDrop = function () {
  const [items, setItems] = useState(initialItems)

  return (
    <div>
      {items.map((item) => {
        return <Item item={item} />
      })}
    </div>
  )
}

const Item = function ({ item }) {
  const r = useMultiDrag({
    item: { type: "card", color: "orange" },
    collect: (monitor) => {
      return {
        isDragging: monitor.isDragging(),
      }
    },
  })
  const [
    [dragProps, ref],
    {
      html5: [html5props, html5Drag],
      touch: [touchProps, touchDrag],
    },
  ] = r
  return (
    <div
      ref={ref}
      className="alert alert-primary"
      key={item.id}
      style={{ opacity: dragProps.isDragging ? 0.5 : 1 }}
    >
      <div ref={html5Drag}>{item.title}</div>
      <div ref={touchDrag}>Handle</div>
    </div>
  )
}

Page.getInitialProps = (context) => {
  return {}
}

export default Page
```

**Comment by thesunny** [comment]
Thank you for fixing the documentation!

**Comment by LouisBrunner** [comment]
The touch worked in my examples, I'll try your use case!
For the handle/container, I am afraid it will be a bit difficult to do, I will look into it though.

**Comment by LouisBrunner** [comment]
The Touch works, you just need to add a `<Preview />` component (https://github.com/LouisBrunner/dnd-multi-backend/tree/master/packages/react-dnd-multi-backend#preview) as the TouchBackend doesn't generate one.

**Comment by LouisBrunner** [comment]
Hi @thesunny,

Is the current implementation usable despite only the handle moving and not the whole container? I don't currently have a workaround for that case, however I'd like to publish the current draft (in order to see the impact and gather more feedback) but I am curious to know if it's at all useful for your original issue.

**Comment by thesunny** [comment]
Hi @LouisBrunner, sorry I dropped the ball on this.

I'll have to check it out again but I think I can answer this more quickly by asking a few questions.

When you drag with the handle on the mouse dragging, the handle can be the container am I correct? In this case, it should work fine since we usually only have a smaller handle within the container when dragging using touch.

When using touch, since you have to define the `<Preview />` component, then I assume touch dragging will show the right thing dragging.

Sorry I'm answering with a bunch of questions, but if my assumptions above are right, I'd say it would work to solve my use case. Mouse dragging is almost always the container, and touch dragging is usually a small handle. In these cases, we get the right preview than it works.

**Comment by LouisBrunner** [comment]
Yes, both of them are right.

You should be able to assign one ref to the container and one to a sub component so the mouse drags the whole thing and touch drags using a smaller, inner handle.

The `<Preview />` component will render based on the associated generator function (as TouchBackend doesn't generate previews because it doesn't use html5 drag'n'drop API).

**Comment by LouisBrunner** [comment]
This will be released as part of 6.1.0
